### ISSUE TITLE
Guardado local

### ISSUE THREAD

**Issue by ifoche** [EyeSeeTea/malariapp#22]
Manejo del guardado local y potencialmente gestión de múltiples formularios

**Comment by josemp10** [comment]
Is this issue referring to meta-data (org. units, periods/event_dates) as well or just data? If it is including meta-data, I am going to add myself to this one. If not, we should create other issue for meta-data,

**Comment by adrianq** [comment]
This is like the overall issue which includes many different task such as:
- [ ] Decide how we are going to deal with metadata information
- [ ] Adapt db to new structure
- [ ] Adapt db for multiple surveys (there is a specific task for this #33) Anyway I am not sure if finally this is needed....
- [ ] Save metadata information
- [ ] Save information

Bottom line, yes maybe we can create an specific issue for the metadata aspects and keep this as the general card for the whole task. I am unassigning this to myself by the moment

**Comment by josemp10** [comment]
Ok, I believe there are 2 major issues here:

1) Dealing with metadata  -> We need to answer both questions: (a) How are we going to get this information? (Most likely calling the DHIS2 API for org. units and periods), (b) when does the user select the metadata? (I hope before entering other data in the surveys). 'Adapt db to new structure' and 'Adapt db for multiple surveys' (if needed) depend very much on the answers to both questions.

2) Dealing with data -> When the user fills a field we should save the value in our local database (which event should we listen? losing the control of the component?). Also we need to be clear about 1) before working on 2)

For me it is fine working on the metadata side...

**Comment by adrianq** [comment]
Agreed.  Let's start with the metadata. Just saw your mail to Rodolfo let's wait for his answer. 
Just some ideas that come to my mind:
- If we are going to retrieve the org.units and periods from DHIS some things need to bare in mind. In order to do this we need to have internet connection that is sth we can't assure in every single case. Therefore, one solution would be to ask DHIS when the app starts and store this information in a table/s that would work as a sort of cache.  If there is not Internet connection we will have to work with the information  that has been previously stored in this cache. This means that the first time you run the app you need to have internet connection otherwise you won't be able to do anything.
- Do the user need to login to the app?
- If  we allow multiple surveys (I know this is sth that needs to be decided yet.) The first screen will be sth like a two-buttons layout saying 1) Begin New Survey 2)Load Survery. If he clicks on load survey a list of surveys (Org unit + period) will be displayed. (I am assuming no user authentication is required otherwise the login panel will be the first on in the app).

As said before let's wait for Rodolfo. Afterwards I think we need to clarify the use case. first of all and once this is done we can think in the db. Maybe when everything is a little bit more clear the three of us can have a design session where we can write down the use case and the db scheme. Yes, I do think that by the moment it is a good idea that @josemp10 leads this... :)

@josemp10 @ifoche Do these questions/ideas make sense?

**Comment by josemp10** [comment]
Yes, you have covered everything. Just some quick thoughs:

- The DHIS2 android app contains a button to sync with a DHIS2 server (in case there is an internet connection). I think we can implement something like that (and likely, as Rodolfo said, reuse part of their code). 

- No idea about login. We can postpone this discussion to the next week. 

- 100% agree with what you said about having multiple surveys. The list of surveys can be implemented using a ListView ;)

**Comment by ifoche** [comment]
I think all the questions are pertinent. 
- It will be important at any time, to think about how to deal with asynchronous logic data corruption. I mean. A user downloads the org units. He goes to the field and fill in the form using a concrete org unit. When he come back to the central and try to upload the data, he detects that org unit has been renamed. As the user cannot come back to the health center and we don't want him to lose all the data he's made the effort of retrieving just because an error. That issue should be taken into account.
- The point of the login I think is crucial. From my point of view, we should prepare the app for being able not only to login, but also to have its loging information synchronized with a DHIS server. Probably in the future it will be desirable to manage the users that can open the app and their permissions from the DHIS server. To avoid future "death refactors" it would be nice to give a look to DHIS login DB structure and whether there is a way of using the same structure. I think is important even if we don't put a login screen yet. Health data is usually covered by a strong protection by law in several countries, so this effort will be surely a must.
- Regarding the the third point, my only comment would be. Depending on whether we chose to prepare ourselves for app login, I would or not do another thing. If we have a login, we probably will have also permissions attached to surveys. I mean, a survey will have to contain any kind of information in the DB indicating who is allowed to do what. As before, prepare ourselves to this may not totally be a pain in the ass, but can save us a lot of future time. Of course, I would use ListView for showing every new thing, as Jose already said :D.

**Comment by josemp10** [comment]
Brief comments to Nacho's notes:

- The metadata sync between DHIS2 and the android app is only going to be in one way (from DHIS2 to the android app). Each org. unit has its own uid (unique identifier) or a code, so the match between the org. units is not going to be based on the name, but the uid.  A user can not edit the uid. I think this keeps the whole thing quite simple.

- About login. Not sure. We don´t have enough information yet. The main point here is to know if different users may have access to different org. units (or not). It really depends on the context. They are not clinical records (no patient information as name, bithdate,..), so the information protection may not be an issue (although I agree it depends on each MoH). Anyway, we have a very nice API in DHIS2 to retrieve information of the users and their roles (although we only are going to have one role), In our app the relation of possible users with org. units and with surveys is still not clear, and we may need to implement other core things first. That was the reason of saying we can leave it for the next week.

- All possible users are going to have same permissions (filling the survey). What we don´t know yet  if they all will have access to all org. units or not...

**Comment by ifoche** [comment]
Ok, I see. Some more comments:
- For the 1st. Yes I didn't meant the user change the ID, but DHIS. If something happens with DHIS org unit IDs, or just imagine that an org unit is still not created in DHIS when the people that is going to fill in the form goes offline, then our app wouldn't do the push. The user must be able to, for example, decide to visit a new health center while he is offline, not being forced to re-enter a connected area. That's the reason why I think we should manage the thing by letting the user change the org unit at the moment when (s)he is uploading the data. A simple "org unit match" would solve the problem. Whenever you're uploading, you should at least be asked like "This org unit you entered, is suposed to be this one from DHIS?" And if you say yes, then you go on. In my ideal there would be also a "No, this is not, and no other option you've provided is, so let me create a new one in this moment". 
- For the 2nd. You know much more about medical information laws than me, so if you say there won't be any problem, that's ok. But I wanted to push the question, just in case we could make a little login effort (that I think could be minimal) and gain a lot. If there's no problem at all, of course it is always better not to write unnecessary code :).
- Yes, all possible users would have the same permission probably, but when you put a user authentication you're not only trying to decide permissions between users of your system. i.e. You may want to avoid that anyone with the app (downloaded using Google Play) uploads data to a server (any malicious user that downloads the app could destroy your service otherwise), or you may want to give a user permission to fill in the form, but not to upload the data, in order to provide another user the rol of supervisor for all the data retrieved by the first one. I don't know, maybe I'm paranoid with this, but when you deal with the data that a lot of people has spent months in retrieving...I think it is important to be sure that nothing bad can happen to this data. I'm probably wrong and nothing will happen, but if it does, somebody will say that our app was not prepared to control that. Isn't it?

**Comment by josemp10** [comment]
Ok, yes, you are probably right with the users. It is necessary to provide a user and a passwd in order to access to the DHIS2 API. But I really think we don´t need to be worried about sync... Anyway too much discussion for me today :), we can have further discussion on Thursday talking about pendings issues and architecture...

**Comment by ifoche** [comment]
For the moment, as login activity is still not implemented, I'm using a dummy user, dummy orgunit and so, a dummy survey. This is all held by a so called "Session" class accessible anywhere. 

Custom tabs are not very well tested, but as they're about to dissappear, it doesn't matter.
